### ISSUE TITLE
Yellow underline with reportUndefinedVariable when creating __eq__ method on dataclass

### ISSUE THREAD

**Issue by blackary** [microsoft/pylance-release#491]
## Environment data

-   Language Server version: 2020.10.0
-   OS and version: OSX 10.15.6 
-   Python version (& distribution if applicable, e.g. Anaconda): 3.8.2

## Expected behaviour

No yellow underline when defining `__eq__` method on a dataclass

## Actual behaviour

Yellow underline appears, with `"__eq__" is not definedPylance (reportUndefinedVariable)`

## Code Snippet / Additional information

```python
from dataclasses import dataclass


@dataclass
class A:
    x: int

    def __eq__(self, other: "A") -> bool:
        return self.x == other.x


class B:
    x: int

    def __eq__(self, other: "B") -> bool:
        return self.x == other.x
```

Yellow underline appears on dataclass class, but not on non-dataclass class.

![image](https://user-images.githubusercontent.com/4040678/96035669-5a884b00-0e31-11eb-86c3-a3af89efa4f2.png)

**Comment by jakebailey** [comment]
I believe this is the same as microsoft/pyright#1098 and should be fixed in our release today.

**Comment by jakebailey** [comment]
Nope, I'm wrong:

![image](https://user-images.githubusercontent.com/5341706/96036852-e7bfa580-0e19-11eb-8938-22385b37582e.png)

**Comment by jakebailey** [comment]
This is related to semantic tokenization, if you set the following the message will go away on a reload.

```
{
	"[python]": {
        "editor.semanticHighlighting.enabled": false,
    }
}
```

@huguesv

**Comment by erictraut** [comment]
@jakebailey, are you able to repro this in pylance with the latest code?

**Comment by jakebailey** [comment]
Yes, that's what the screenshots are of above, just our code at the latest.

I have a "fix" that just suppresses all diagnostics during the semantic token walk which I think would solve this once and for all, so long as the semantic token pass isn't produce diagnostics that the background isn't.

**Comment by erictraut** [comment]
Yeah, that's not the right fix. That will simply suppress what is a real bug. 

This is likely a real bug in the type evaluator triggered by a particular order of evaluation that the semantic token code is using. If I can find a repro, I should be able to fix it in short order. I wasn't able to repro the other similar bug with the latest code.

**Comment by jakebailey** [comment]
IIRC we disable the background thread on debug (because node doesn't make it easy to debug it), and that's where the diagnostics / semantic tokenization happens. That probably doesn't help in trying to find the issue. 🙁

**Comment by erictraut** [comment]
Yeah, I temporarily disable the background analysis when I'm debugging issues like this.

**Comment by jakebailey** [comment]
Oh, what am I saying, I ran this in debug mode (I didn't make a VSIX or anything), so there's at least some hope.

**Comment by erictraut** [comment]
This was a bug in the semantic token provider. It was calling the type evaluator to evaluate the type of an expression, but it wasn't passing an expression node. I've made the type evaluator more resilient so it can handle this case. This was an easier and more robust way to fix this than finding all the places where the caller was passing a non-expression node.

The fix will be in the next release of pylance.

**Comment by jakebailey** [comment]
This issue has been fixed in version 2020.10.2, which we've just released. You can find the changelog here: https://github.com/microsoft/pylance-release/blob/master/CHANGELOG.md#2020102-21-october-2020
